File: phases/00-setup-and-tooling/04-apis-and-keys/code/first_api_call_gemini.ts
// Phase 0 · Lesson 04 — APIs and keys (Gemini TypeScript port).
// Reads GEMINI_API_KEY from env, then makes one generateContent call using
// both the official SDK and raw HTTP. Set MOCK=1 to skip the network.
// Refs: https://ai.google.dev/gemini-api/docs/quickstart
//       https://nodejs.org/api/globals.html#fetch (Node 18+ ships fetch)

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import process from "node:process";

type GenerateContentResponse = {
  candidates: { content: { parts: { text: string }[] } }[];
  usageMetadata?: { promptTokenCount: number; candidatesTokenCount: number };
};

const MOCK_RESPONSE: GenerateContentResponse = {
  candidates: [
    {
      content: {
        parts: [
          {
            text: "A neural network is a stack of differentiable functions that learns patterns by adjusting weights against a loss signal.",
          },
        ],
      },
    },
  ],
  usageMetadata: { promptTokenCount: 12, candidatesTokenCount: 28 },
};

// .env loader — same pattern as first_api_call.ts. KEY=VALUE per line, # comments.
function loadDotenv(path: string): Record<string, string> {
  let raw: string;
  try {
    raw = readFileSync(path, "utf8");
  } catch {
    return {};
  }
  const out: Record<string, string> = {};
  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq <= 0) continue;
    const key = trimmed.slice(0, eq).trim();
    let value = trimmed.slice(eq + 1).trim();
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }
    out[key] = value;
  }
  return out;
}

function mergeEnv(): Record<string, string | undefined> {
  const fromFile = loadDotenv(resolve(process.cwd(), ".env"));
  return { ...fromFile, ...process.env };
}

function buildGeminiUrl(model: string): string {
  // The full REST endpoint. SDKs build this URL internally.
  return `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent`;
}

function buildGeminiBody(prompt: string): string {
  // The JSON payload — manually constructed here, built by the SDK in callWithSdk.
  return JSON.stringify({
    contents: [{ parts: [{ text: prompt }] }],
  });
}

/** Make an API call using the official Google Gen AI SDK (@google/genai).
 *
 * The SDK handles: authentication, HTTP connection, request formatting,
 * response parsing, error handling, and retries — all automatically. */
async function callWithSdk(
  apiKey: string,
  prompt: string,
  model: string,
): Promise<number> {
  if (process.env.MOCK === "1" || apiKey === "mock") {
    const text = MOCK_RESPONSE.candidates[0].content.parts[0].text;
    process.stdout.write(`SDK response: ${text}\n`);
    return 0;
  }
  try {
    // Dynamic import so the script doesn't crash if the SDK isn't installed.
    const { GoogleGenAI } = await import("@google/genai");
    const client = new GoogleGenAI({ apiKey });
    const response = await client.models.generateContent({
      model,
      contents: prompt,
    });
    // response.text gives the model's reply directly — no JSON parsing needed.
    const text = response.text ?? "";
    process.stdout.write(`SDK response: ${text}\n`);
    if (response.usageMetadata) {
      process.stdout.write(
        `Tokens used: ${response.usageMetadata.promptTokenCount} in, ` +
          `${response.usageMetadata.candidatesTokenCount} out\n`,
      );
    }
    return 0;
  } catch (err) {
    process.stderr.write(`SDK call failed: ${(err as Error).message}\n`);
    return 1;
  }
}

/** Make the same API call using raw HTTP fetch — no SDK, no third-party imports.
 *
 * This is what the SDK does internally. Seeing it demystifies the "magic". */
async function callRawHttp(
  apiKey: string,
  prompt: string,
  model: string,
): Promise<number> {
  if (process.env.MOCK === "1" || apiKey === "mock") {
    // Use the fixture so MOCK=1 returns the same shape as a real response.
    const text = MOCK_RESPONSE.candidates[0].content.parts[0].text;
    process.stdout.write(`Raw HTTP response: ${text}\n`);
    return 0;
  }

  try {
    const url = buildGeminiUrl(model);
    // HTTP headers — we must set Content-Type ourselves; API key goes as a query param.
    const headers = { "Content-Type": "application/json" };
    // Build and encode the request body manually
    const body = buildGeminiBody(prompt);

    // Fetch is built into Node 18+. This is exactly what the SDK wraps.
    const resp = await fetch(`${url}?key=${apiKey}`, {
      method: "POST",
      headers,
      body,
    });

    if (!resp.ok) {
      const bodyText = await resp.text();
      throw new Error(`Gemini ${resp.status}: ${bodyText.slice(0, 200)}`);
    }
    // Parse the raw JSON response into a typed object
    const result = (await resp.json()) as GenerateContentResponse;
    // Navigate nested objects to extract the generated text
    const text = result.candidates[0].content.parts[0].text;
    process.stdout.write(`Raw HTTP response: ${text}\n`);
    if (result.usageMetadata) {
      process.stdout.write(
        `Tokens used: ${result.usageMetadata.promptTokenCount} in, ` +
          `${result.usageMetadata.candidatesTokenCount} out\n`,
      );
    }
    return 0;
  } catch (err) {
    process.stderr.write(`Raw HTTP call failed: ${(err as Error).message}\n`);
    return 1;
  }
}

async function main(): Promise<number> {
  const env = mergeEnv();
  const apiKey = env.GEMINI_API_KEY ?? "mock";
  const usingMock = process.env.MOCK === "1" || apiKey === "mock";
  const model = "gemini-2.0-flash-lite";
  const prompt = "What is a neural network in one sentence?";

  process.stdout.write("=== Gemini API Calls ===\n\n");
  process.stdout.write(
    usingMock
      ? "Mode: MOCK (no network). Unset MOCK and export GEMINI_API_KEY for a live call.\n\n"
      : "Mode: LIVE.\n\n",
  );

  process.stdout.write("1. Using the SDK:\n");
  const sdkCode = await callWithSdk(apiKey, prompt, model);

  process.stdout.write("\n2. Using raw HTTP:\n");
  const httpCode = await callRawHttp(apiKey, prompt, model);

  return sdkCode === 0 && httpCode === 0 ? 0 : 1;
}

main().then((code) => process.exit(code));

File: phases/00-setup-and-tooling/04-apis-and-keys/code/first_api_call_gemini.py
import os       # Access environment variables (like GEMINI_API_KEY)
import json      # Convert Python dicts to JSON strings and parse JSON responses
import urllib.request  # Built-in HTTP client — no external dependencies needed for raw requests


def call_with_sdk():
    """Make an API call using Google's official Python SDK (google-genai).
    
    The SDK handles: authentication, HTTP connection, request formatting,
    response parsing, error handling, and retries — all automatically.
    """
    try:
        from google import genai   # Third-party SDK from Google
    except ImportError:
        print("Install the SDK: pip install google-genai")
        return

    # Create a client — reads the API key from the argument (or env var GOOGLE_API_KEY by default)
    client = genai.Client(api_key=os.environ.get("GEMINI_API_KEY"))
    # .generate_content() is a high-level method — it builds the HTTP request,
    # sends it, parses the response, and returns a Python object
    response = client.models.generate_content(
        model="gemini-2.0-flash-lite",
        contents="What is a neural network in one sentence?",
    )
    # response.text gives the model's reply directly — no JSON parsing needed
    print(f"SDK response: {response.text}")
    # usage_metadata has token counts — the SDK extracts them from the raw response
    usage = response.usage_metadata
    if usage:
        print(f"Tokens used: {usage.prompt_token_count} in, {usage.candidates_token_count} out")
    # The SDK also handles: authentication headers, retries on 429/503, streaming, etc.


def call_raw_http():
    """Make the same API call using raw HTTP — no SDK, no third-party imports.
    
    This is what the SDK does internally. Seeing it demystifies the "magic".
    """
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("Set GEMINI_API_KEY environment variable first")
        return

    model = "gemini-2.0-flash-lite"
    # The full REST endpoint URL. Every SDK call maps to one of these.
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent"
    # HTTP headers — we must set Content-Type ourselves
    headers = {"Content-Type": "application/json"}
    # Build the request body manually as a JSON string, then encode to bytes
    body = json.dumps({
        "contents": [{"parts": [{"text": "What is a neural network in one sentence?"}]}]
    }).encode()

    # Create a POST request object with URL, body, headers, and method
    req = urllib.request.Request(
        url, data=body, headers=headers, method="POST"
    )
    # Send the request and read the response
    with urllib.request.urlopen(req) as resp:
        # Parse the raw JSON response into a Python dict
        result = json.loads(resp.read())
        # Navigate nested dicts to extract the generated text
        text = result["candidates"][0]["content"]["parts"][0]["text"]
        print(f"Raw HTTP response: {text}")
    # Notice: we didn't handle auth headers, retries, or error formatting here.
    # The SDK does all of that for you.


if __name__ == "__main__":
    print("=== Gemini API Calls ===\n")
    print("1. Using the SDK:")
    call_with_sdk()
    print("\n2. Using raw HTTP:")
    call_raw_http()
